### PR TITLE
feat(logging): RHICOMPL-1423 AuditLog wrapper

### PR DESCRIPTION
**Pull request by vkrizan**
Reworks the AuditLog to provide audit capabilities by wrapping a base
logger.  This allows to configure any logger (e.g. Rails.logger) to have
an audit method.

The AuditLog Middleware requires the `Rails.logger` to be wapped,
otherwise it falls back to an info logging.

Audit logs can be sent through `audit_success` or `audit_fail` depending on the status.

Example Rails configuration:
```
  # Rails config
  # base_logger = ActiveSupport::Logger.new(STDOUT)
  config.logger = Insights::API::Common::AuditLog.new(base_logger, Logger.new('log/myaudit.log'))
  config.middleware.use Insights::API::Common::AuditLog::Middleware
```

Example usage:
```
  Rails.logger.audit('Audit message')
  Rails.logger.audit({ controller: 'MyCtrl', message: 'with evidence' })
  Rails.logger.audit_success('Success!')
  Rails.logger.audit_fail('Fail!')

  # setting account number evience context (with optional block)
  Insights::API::Common::AuditLog.audit_with_account('123')
  # or
  Rails.logger.audit_with_account('123')
```

(Apologies for too many changes.)

## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [X] Input Validation
- [X] Output Encoding
- [X] Authentication and Password Management
- [X] Session Management
- [X] Access Control
- [X] Cryptographic Practices
- [X] Error Handling and Logging
- [X] Data Protection
- [X] Communication Security
- [X] System Configuration
- [X] Database Security
- [X] File Management
- [X] Memory Management
- [X] General Coding Practices

### CODE REVIEW

File: test/lib/audit_log/audit_log_test.rb [hunk 1]
```diff
@@ -4,59 +4,31 @@
...
+    assert wrapped.respond_to?(:audit_fail)
   end
 
-  def capture_log(msg)
-    @audit.logger.info(msg)
-    @output.rewind # logger seems to read what it's writing?
-    JSON.parse @output.readlines[-1]
-  end
-
-  test 'logs a general message formatted into JSON' do
```
Comment (vkrizan) [inline test/lib/audit_log/audit_log_test.rb:19]: Moved to `wrapped_logger_test.rb`

**Comment by vkrizan** [comment]
"codecov/patch — 97.01% of diff hit (target 97.42%)", almost! :D

**Review by akofink — APPROVED** [review]
Nice work, @vkrizan! I have still the same comment as before, that IMO we should go ahead and enable some version of this (i.e. log/audit.log) by default in devel so it's easy to see it working and easy to debug any issues once we enable it in deployed environments. For example, in `config/development.rb`:

```rb
  # Use our custom audit logger
  require 'audit_log/audit_log'
  base_logger = ActiveSupport::Logger.new(STDOUT)
  config.logger = Insights::API::Common::AuditLog.new(base_logger, Logger.new('log/audit.log'))
  config.middleware.use Insights::API::Common::AuditLog::Middleware
```

Logs:
```js
{"@timestamp":"2021-02-12T22:35:39.520955Z","hostname":"6b9e27602833","pid":1,"thread_id":"2b286f080c60","level":"audit","transaction_id":"39a0246659129095fe96","message":"POST /api/compliance/graphql -> 200 OK","status":"success","controller":"GraphqlController#query","remote_ip":"127.0.0.1"}
{"@timestamp":"2021-02-12T22:35:40.578300Z","hostname":"6b9e27602833","pid":1,"thread_id":"2b286f080a58","level":"audit","transaction_id":"8ec51f0d6eafd28d5a64","message":"POST /api/compliance/graphql -> 200 OK","status":"success","controller":"GraphqlController#query","remote_ip":"127.0.0.1"}
{"@timestamp":"2021-02-12T22:35:40.998003Z","hostname":"6b9e27602833","pid":1,"thread_id":"2b286f080d78","level":"audit","transaction_id":"8a59c0d951d5d99d4cea","message":"POST /api/compliance/graphql -> 200 OK","status":"success","controller":"GraphqlController#query","remote_ip":"127.0.0.1"}
{"@timestamp":"2021-02-12T22:35:55.473050Z","hostname":"6b9e27602833","pid":1,"thread_id":"2b286f080c60","level":"audit","transaction_id":"39a0246659129095fe96","message":"GET /api/compliance/v1/systems -> 200 OK","status":"success","controller":"V1::SystemsController#index","remote_ip":"127.0.0.1"}
{"@timestamp":"2021-02-12T22:36:27.298962Z","hostname":"6b9e27602833","pid":1,"thread_id":"2b286f08097c","level":"audit","transaction_id":"c31fc998d3d871f428d7","message":"GET /api/compliance/v1/profiles -> 200 OK","status":"success","controller":"V1::ProfilesController#index","remote_ip":"127.0.0.1"}
```

**Comment by vkrizan** [comment]
@akofink Make sense. I just wanted to avoid any hardcoded log path, esp. if they are being stored on SSD disk and not in temp. And also avoid this:
```
1,7G	development.log
4,0G	test.log
```

**Comment by akofink** [comment]
My `log` directory is 28G :) It's never been an issue before, and it's easy to truncate. We could also have the config there but commented out in development.rb. Either way, this PR is good to merge. Thanks again!

**Comment by vkrizan** [comment]
@akofink added three new commits, last one adds the logging on dev. Please check/try.

**Review by akofink — APPROVED** [review]
Works as expected. Thanks, @vkrizan!

**Pull request merged**